projecttacoma/deqm-test-server
# New import

# Summary
This PR changes the workflow of the bulk $import operation. Previously, this server used the Ping and Pull approach, but now it uses the proposed bulk data import outlined in [this IG](file:///Users/eperelli/Downloads/proposed-bulk-import-ig/site/index.html). 

## New behavior
For more information on the previous Ping and Pull approach, refer to this repository's Wiki page [here](https://github.com/projecttacoma/deqm-test-server/wiki/Bulk-Data-Import). The main difference between that approach and the new approach is that the new approach utilizes already exported inputs rather than this server kicking off the export request. This simplifies the process by including the export ndjson urls in the Import Manifest in the request body of the $import operation rather than just the export url. 

## Code changes
- `package.json` / `package-lock.json` - removed `bulk-data-utilities` dependency.
- `dbOperations.js` - two new functions for storing and getting ndjson file failed outcomes, inclusion of the import manifest in the `bulkImportClient` object so that it may be included on the Import Result.
- `importWorker.js` - `executeNewImportWorkflow` replaces `executePingAndPull`.
- `ndjsonWorker.js` - ndjson failed outcomes are now stored as well.
- `bulkstatus.service.js` - **WORK IN PROGRESS** - the structure of the Import Result object has yet to be finalized.
- `import.service.js` - use `retrieveInputUrls`.
- `measure.service.js` - `bulkSubmitData` now follows the new `$import` workflow. This is okay for now because `$bulk-submit-data` may even go away.
- `exportUtils.js` - `retrieveInputUrls` now returns all ndjson fileUrls included in the Import Manifest
- `test` - deleted param files that are no longer used in tests, updated unit tests to follow new workflow

# Testing guidance
- `npm run check`
- Utilize the [bulk-export-server](https://github.com/projecttacoma/bulk-export-server) to get the export ndjson url files to include in the Import Manifest. This can be done by populating the bulk-export-server, kicking off the $export request, and then polling the bulkstatus location until ndjson file urls are returned. Make sure the `bulk-export-server` is running on a different port than `deqm-test-server` (`PORT=3001 npm start`).
- Spin up the `deqm-test-server`: `npm start`
- Send a GET request to `http://localhost:3000/4_0_1/$import` with an Import Manifest json object in the request body (defined in IG) that includes ndjson file urls from the `bulk-export-server`. I am happy to send examples, but I figured leaving this open would be better for testing to make sure that I interpreted the IG correctly.
- Make sure the $import kickoff returns a 202 status.
- Send a GET request to the content-location (bulkstatus polling location). Make sure this either returns in-progress or OK status with a Import Result object.

# Import Results
Note that the structure of the Import Results (`bulkstatus.service.js`) is still a work in progress.
